### PR TITLE
ASTTableIdentifier Part #1: improve internal representation of ASTIdentifier name

### PR DESCRIPTION
**Pull request by abyss7**
I hereby agree to the terms of the CLA available at: https://yandex.ru/legal/cla/?lang=en

Changelog category (leave one):
- Other

Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Use only |name_parts| as primary name source and auto-generate full name

### CODE REVIEW

**Comment by abyss7** [comment]
Yandex third-party fixed

File: src/Interpreters/TranslateQualifiedNamesVisitor.cpp [hunk 1]
```diff
@@ -175,9 +175,12 @@ void TranslateQualifiedNamesMatcher::visit(ASTSelectQuery & select, const ASTPtr
 
 static void addIdentifier(ASTs & nodes, const DatabaseAndTableWithAlias & table, const String & column_name)
 {
+    std::vector<String> parts = {column_name};
+
```
Comment (tavplubix) [inline src/Interpreters/TranslateQualifiedNamesVisitor.cpp:179]:
```suggestion
    std::vector<String> parts;
```

File: src/Interpreters/TranslateQualifiedNamesVisitor.cpp [hunk 2]
```diff
@@ -175,9 +175,12 @@ void TranslateQualifiedNamesMatcher::visit(ASTSelectQuery & select, const ASTPtr
...
     String table_name = table.getQualifiedNamePrefix(false);
-    auto identifier = std::make_shared<ASTIdentifier>(std::vector<String>{table_name, column_name});
-    nodes.emplace_back(identifier);
+    if (!table_name.empty()) parts.insert(parts.begin(), table_name);
+
```
Comment (tavplubix) [inline src/Interpreters/TranslateQualifiedNamesVisitor.cpp:182]:
```suggestion
    if (!table_name.empty())
        parts.push_back(std::move(table_name));
    parts.push_back(column_name);
```

File: src/Parsers/ASTIdentifier.cpp [hunk 3]
```diff
@@ -16,59 +16,65 @@ namespace ErrorCodes
...
+    : name_parts(name_parts_), semantic(std::make_shared<IdentifierSemanticImpl>())
 {
-    auto ret = std::make_shared<ASTIdentifier>(*this);
-    ret->semantic = std::make_shared<IdentifierSemanticImpl>(*ret->semantic);
-    return ret;
+    assert(!name_parts.empty());
+    for (const auto & part : name_parts)
+    {
+        assert(!part.empty());
+        (void) part;  // otherwise not-used in release build
```
Comment (tavplubix) [inline src/Parsers/ASTIdentifier.cpp:32]: Does `[[maybe_unused]]` work with `for`?

File: src/Parsers/ASTIdentifier.cpp [hunk 4]
```diff
@@ -16,59 +16,65 @@ namespace ErrorCodes
...
-    if (name_parts.empty())
-        return;
-    name = name_parts[0];
-    for (size_t i = 1; i < name_parts.size(); ++i)
-        name += '.' + name_parts[i];
+    assert(!name_parts.empty());
+
+    if (full_name.empty())
+    {
+        full_name = name_parts[0];
```
Comment (tavplubix) [inline src/Parsers/ASTIdentifier.cpp:72]: Do I understand correctly, that it cannot be called from multiple threads (or, at least, `full_name` is initialized on early stages of query processing when there is only one thread)?

File: tests/queries/server.py [hunk 5]
```diff
@@ -137,19 +137,34 @@ def stop(self):
...
-                 </replica>
-             </shard>
-             <shard>
-                 <replica>
-                     <host>localhost</host>
-                     <port>{tcp_port}</port>
-                 </replica>
-             </shard>
-         </test_cluster_two_shards_localhost>
+            <shard>
```
Comment (tavplubix) [inline tests/queries/server.py:140]: What is the purpose of this change?

**Review by tavplubix — APPROVED** [review]

File: tests/queries/server.py (hunk 5 above)
Comment (abyss7) [reply tests/queries/server.py:140]: To allow more tests to pass - using pytest framework. Without this section of config they don't work.

File: src/Parsers/ASTIdentifier.cpp (hunk 4 above)
Comment (abyss7) [reply src/Parsers/ASTIdentifier.cpp:72]: It's a good point to consider - I will recheck the name initialization

File: src/Parsers/ASTIdentifier.cpp (hunk 3 above)
Comment (abyss7) [reply src/Parsers/ASTIdentifier.cpp:32]: Will check.
